src/Orleans/Configuration/New/ClientFactoryBuilder.cs

```diff
@@ -0,0 +1,309 @@
+        /// </summary>
+        public ClientConfiguration.GatewayProviderType GatewayProvider { get; set; }
+
+        private static readonly TimeSpan DEFAULT_GATEWAY_LIST_REFRESH_PERIOD = TimeSpan.FromMinutes(1);
+
+        /// <summary>
+        /// </summary>
+        public bool UseAzureSystemStore 
+        { 
+            get { 
+                return GatewayProvider == Configuration.ClientConfiguration.GatewayProviderType.AzureTable 
+                       && !String.IsNullOrWhiteSpace(Settings.SystemStore.DeploymentId) 
+                       && !String.IsNullOrWhiteSpace(Settings.SystemStore.DataConnectionString); 
+            } 
+        }
+
+        /// <summary>
+        /// </summary>
+        public bool UseSqlSystemStore
+        {
+            get
+            {
+                return GatewayProvider == Configuration.ClientConfiguration.GatewayProviderType.SqlServer
+                && !String.IsNullOrWhiteSpace(Settings.SystemStore.DeploymentId)
+                && !String.IsNullOrWhiteSpace(Settings.SystemStore.DataConnectionString);
+            }
+        }
+
+        private bool HasStaticGateways { get { return Gateways != null && Gateways.Count > 0; } }
+        /// <summary>
+        /// </summary>
+        public IDictionary<string, ProviderCategoryConfiguration> ProviderConfigurations { get; set; }
+        
+        /// <summary>
+        /// </summary>
+        public ClientFactoryBuilder()
+        {
+            creationTimestamp = DateTime.UtcNow;
+            SourceFile = null;
+            PreferedGatewayIndex = -1;
+            Gateways = new List<IPEndPoint>();
+            DNSHostName = Dns.GetHostName();
+
+            GatewayListRefreshPeriod = DEFAULT_GATEWAY_LIST_REFRESH_PERIOD;
+            LimitManager = new LimitManager();
+            ProviderConfigurations = new Dictionary<string, ProviderCategoryConfiguration>();
+        }
+        
+        
+        public ClientFactorySettings Settings { get; private set; }
+
+        public void Build(ClientFactorySettings client)
+        {
+            Settings = client;
+
+            foreach(var gateway in Settings.Gateways)
+            {
+                Gateways.Add(gateway.ParseIPEndPoint().GetResult());
+                if (GatewayProvider == ClientConfiguration.GatewayProviderType.None)
+                {
+                    GatewayProvider = ClientConfiguration.GatewayProviderType.Config;
+                }
+            }
+                        
+            if (Settings.SystemStore.SystemStoreType.HasValue)
+            {
+                GatewayProvider = Settings.SystemStore.SystemStoreType.Value;
+            }
+            if (!string.IsNullOrEmpty(Settings.SystemStore.CustomGatewayProviderAssemblyName))
+            {
+                if (Settings.SystemStore.CustomGatewayProviderAssemblyName.EndsWith(".dll"))
+                    throw new FormatException("Use fully qualified assembly name for \"CustomGatewayProviderAssemblyName\"");
```

Comment:
This exception message is confusing. do you want it to use fully qualified assembly name or it shouldn't?